sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/DiagnosticAnalyzerContextHelper.cs

```diff
@@ -191,9 +176,21 @@ internal static class DiagnosticAnalyzerContextHelper
         private static readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<SyntaxTree, bool>> Cache
             = new ConditionalWeakTable<Compilation, ConcurrentDictionary<SyntaxTree, bool>>();
 
-        public static bool ShouldAnalyze(GeneratedCodeRecognizer generatedCodeRecognizer, SyntaxTree syntaxTree, Compilation c, AnalyzerOptions options) =>
-            !syntaxTree.IsGenerated(generatedCodeRecognizer, c) ||
-            options.ShouldAnalyzeGeneratedCode(c.Language);
+        private static bool ShouldAnalyze(SonarAnalysisContext context, GeneratedCodeRecognizer generatedCodeRecognizer, SyntaxTree syntaxTree, Compilation c, AnalyzerOptions options) =>
+            context.ShouldAnalyzeGenerated(c, options) ||
+            !syntaxTree.IsGenerated(generatedCodeRecognizer, c);
+
+        private static bool ShouldAnalyze(CompilationStartAnalysisContext context, GeneratedCodeRecognizer generatedCodeRecognizer, SyntaxTree syntaxTree, Compilation c, AnalyzerOptions options) =>

```

Comment:
I get a roslyn warning on this one (`'analysisContext' does not register any analyzer actions. Consider moving actions registered in 'Initialize' that depend on this start action to 'analysisContext'.`) but I don't really understand why